python/ray/train/bagua.py

```diff
@@ -0,0 +1,184 @@
+from dataclasses import dataclass
+import logging
+import os
+
+from typing import Optional
+
+import ray
+from ray.train.torch import (
+    TorchAccelerator,
+    TorchBackend,
+)
+
+from ray.train.backend import BackendConfig
+from ray.train.session import get_accelerator, set_accelerator
+from ray.train.worker_group import WorkerGroup
+from ray.train.utils import get_address_and_port
+from ray.util import PublicAPI
+
+import torch
+import torch.distributed as dist
+from torch.utils.data import DistributedSampler
+
+import bagua.torch_api
+from bagua.torch_api.algorithms import gradient_allreduce
+
+try:
+    from torch.profiler import profile
+except ImportError:
+    profile = None
+
+logger = logging.getLogger(__name__)
+
+
+def _get_torch_distributed_sampler(dataset, shuffle):
+    return DistributedSampler(
+        dataset,
+        num_replicas=bagua.torch_api.get_world_size(),  # equivalent to ray.train.size()
+        rank=bagua.torch_api.get_rank(),  # equivalent to ray.train.world_rank()
+        shuffle=shuffle,
+    )
+
+
+class BaguaAccelerator(TorchAccelerator):
+    def __init__(self, amp: bool = False):
+        super().__init__(amp=amp)
+
+    def prepare_model(
+        self,
+        model: torch.nn.Module,
+        optimizer: torch.optim.Optimizer,
+        algorithm: bagua.torch_api.algorithms.Algorithm,
+    ) -> torch.nn.Module:
+
+        device = self.get_device()
+        model = model.to(device)
+
+        if torch.cuda.is_available():
+            torch.cuda.set_device(device)
+
+        if self.amp_is_enabled:
+            # move wrap_forward() and model_get_state() to a new function
+            model = self._patch_model_forward_and_state(model)
+
+        logger.info("Wrapping provided model in BaguaDDP.")
+        # we need the optimizer when preparing the model
+        # with_bagua() returns a BaguaModule not torch.nn.Module
+        model = model.with_bagua([optimizer], algorithm)
+
+        return model
+
+
+@PublicAPI(stability="beta")
+@dataclass
+class BaguaConfig(BackendConfig):
+
+    store: Optional[torch.distributed.Store] = None
+
+    @property
+    def backend_cls(self):
+        return BaguaBackend
+
+
+def setup_torch_process_group(store: Optional[torch.distributed.Store] = None):
+    torch.cuda.set_device(bagua.torch_api.get_local_rank())
+    # init_process_group() is different from ray.train.torch approach
+    # https://github.com/BaguaSys/bagua/blob/master/bagua/torch_api/communication.py#L524-L529
+    # 1. backend is nccl in Bagua
+    # 2. init_method is the default value provided in torch
+    # 3. users can only set the store argument
+    bagua.torch_api.init_process_group(store)
```

Comment:
We could implement a customized version of `bagua.torch_api.init_process_group()` to allow other arguments. But it seems to me that it may be safer to use the provided `init_process_group()`. According to this [comment](https://github.com/BaguaSys/bagua/blob/master/bagua/torch_api/communication.py#L522), bagua may remove the dependence on torch process group later.

Reply:
I think you might be able to use this https://bagua.readthedocs.io/en/latest/autoapi/bagua/torch_api/data_parallel/distributed/index.html#bagua.torch_api.data_parallel.distributed.to_bagua_process_group to convert the torch process group to a bagua process group?

Reply:
Thanks for the reference! I was not aware of this handy function :+1: